mono/tests/test-runner.cs

```diff
@@ -154,182 +143,145 @@ class TestInfo {
-					fs.WriteLine (e.Data);
-					fs.Flush ();
-				}
-			};
-
-			p.ErrorDataReceived += delegate (object sender, DataReceivedEventArgs e) {
-				Process p2 = (Process)sender;
-
-				StreamWriter fs;
-
-				lock (monitor) {
-					fs = process_data [p2].stderr;
-
-					if (e.Data == null)
-						process_data [p2].stderr = null;
-
-				}
+		List<Thread> threads = new List<Thread> (concurrency);
 
-				if (e.Data == null) {
-					fs.Close ();
+		for (int j = 0; j < concurrency; ++j) {
+			Thread thread = new Thread (() => {
+				while (true) {
+					TestInfo ti;
 
 					lock (monitor) {
-						process_data [p2].stderr = null;
+						if (test_info.Count == 0)
+							break;
+						ti = test_info.Dequeue ();
 					}
-				} else {
-					fs.WriteLine (e.Data);
-					fs.Flush ();
-				}
-			};
 
-			lock (monitor) {
-				processes.Add (p);
-				process_data [p] = data;
-			}
-			p.Start ();
+					string test = ti.test;
+					string opt_set = ti.opt_set;
+
+					if (concurrency == 1)
+						Console.Write ("Testing " + test + "... ");
 
-			p.BeginOutputReadLine ();
-			p.BeginErrorReadLine ();
+					/* Spawn a new process */
+					string process_args;
+					if (opt_set == null)
+						process_args = test;
+					else
+						process_args = "-O=" + opt_set + " " + test;
+					ProcessStartInfo info = new ProcessStartInfo (runtime, process_args);
+					info.UseShellExecute = false;
+					info.RedirectStandardOutput = true;
+					info.RedirectStandardError = true;
+					Process p = new Process ();
+					p.StartInfo = info;
+
+					ProcessData data = new ProcessData ();
+					data.test = test;
+
+					string log_prefix = "";
+					if (opt_set != null)
+						log_prefix = "." + opt_set.Replace ("-", "no").Replace (",", "_");
+
+					data.stdoutFile = test + log_prefix + ".stdout";
+					data.stdout = new StreamWriter (new FileStream (data.stdoutFile, FileMode.Create));
+
+					data.stderrFile = test + log_prefix + ".stderr";
+					data.stderr = new StreamWriter (new FileStream (data.stderrFile, FileMode.Create));
+
+					p.OutputDataReceived += delegate (object sender, DataReceivedEventArgs e) {
+						if (e.Data != null) {
+							data.stdout.WriteLine (e.Data);
+						} else {
+							data.stdout.Flush ();
+							data.stdout.Close ();
+						}
+					};
+
+					p.ErrorDataReceived += delegate (object sender, DataReceivedEventArgs e) {
+						if (e.Data != null) {
+							data.stderr.WriteLine (e.Data);
+						} else {
+							data.stderr.Flush ();
+							data.stderr.Close ();
+						}
+					};
+
+					p.Start ();
+
+					p.BeginOutputReadLine ();
+					p.BeginErrorReadLine ();
+
+					if (!p.WaitForExit (timeout * 1000)) {
+						lock (monitor) {
+							timedout.Add (data);
+							failed.Add (data);
+						}
 
-			ThreadPool.QueueUserWorkItem (o => {
-				Process process = (Process) o;
+						if (concurrency == 1)
+							Console.WriteLine ("timed out.");
+						else
+							Console.Write (".");
 
-				process.WaitForExit ();
+						p.Kill ();
```

Comment:
Since `Kill` is asynchronous, should we `.WaitForExit()` again to make sure the process has gone or do we not care if it sticks around?

We should probably also wrap it in try/catch to make sure any exception doesn't take the process down.